data/ai/lua/ai_helper.lua

```diff
@@ -639,6 +639,26 @@ function ai_helper.is_opposite_adjacent(hex1, hex2, center_hex)
     return false
 end
 
+function ai_helper.get_locations_no_borders(location_filter)
```

Comment:
iirc locations filters also have an attribute include_borders=yes(default)/no which could be even faster.

Reply:
Note though that attribute include_borders=yes(default)/no  is iirc afaik recursiveley given su subfilters, so a filter with include_borders=no with also ignore border tiles when doing a [filter_adjacent] check.

Reply:
Hmm, good point.  That's new in 1.13, so I didn't think of it ...
The fact that it also applies to [filter_adjacent] is probably not a problem for the AIs where it is currently needed, but a note to that effect should be added to the ai_helper function.
I'll do some speed tests tonight and report back.

Reply:
> I'll do some speed tests tonight and report back.

Note that it will of course depend on how exactly you arrange them, meaning code like:
```
local filter =  ...
filter.include_borders = false
wesnoth.get_locations(filter)
```
will of course be faster then for example 
```
local filter =  ...
filter.include_borders = false
wesnoth.get_locations({
  include_borders = false,
  { "and", filter }
})
```

Reply:
I did speed tests for this and can confirm what @gfgtdf wrote.  The second method in his post above requires the time for evaluating the filter **plus** getting all hexes of the entire map and excluding hexes, so it can be really slow.  The first method only goes over the filter once and excludes borders from those.  It is much faster than the second method.

By comparison, what I had (evaluating the filter and then running a Lua loop over that) is *a little* slower than @gfgtdf's first method, as it goes over a slightly larger number of hexes (filter incl. border hexes) and because Lua is a bit slower than C++.

Based on this, I suggest modifying the ai_helper function to this (it is important that the input filter table is not modified):
```
function ai_helper.get_locations_no_borders(location_filter)
    -- Returns the same locations array as wesnoth.get_locations(location_filter),
    -- but excluding hexes on the map border.
    --
    -- This is faster than alternative methods, at least with the current 
    -- implementation of standard location filter evaluation by the engine.

    local old_include_borders = location_filter.include_borders
    location_filter.include_borders = false
    local locs = wesnoth.get_locations(location_filter)
    location_filter.include_borders = old_include_borders
    return locs
end
```

Reply:
Note that this code will probably not work if location_filter is a vconfig object, not sure who calls that function so it might not be an issue.

Reply:
Yeah, it's not an issue here, I have now tested it for all uses of the function by mainline Lua AIs. I added your comment to the function comments though.

I've committed and squashed this now, let me know if you have any further comments.